### DISCUSSION TITLE
Migrations with advisory locks

### DISCUSSION THREAD

**Discussion by cavalle** [jeremyevans/sequel#2089] [General]
By default, [ActiveRecord uses advisory locks](https://github.com/rails/rails/blob/v7.1.1/activerecord/lib/active_record/migration.rb#L1436) (if supported by the db adapter, e.g. postgres) to ensure concurrent migrations don’t end up running on the same database. We use Sequel and recently came across the need for such a feature to avoid issues we were having with our deployment setup. We finally implemented a solution following the same approach as Rails. It looks like this:

```ruby
def with_advisory_lock(lock_name)
  lock_id = Zlib.crc32(lock_name.to_s) % MAX_ALLOWED_ADVISORY_LOCK_ID

  got_lock = DB.get(Sequel.function(:pg_try_advisory_lock, lock_id))
  raise AdvisoryLockError unless got_lock

  yield
ensure
  if got_lock && !DB.get(Sequel.function(:pg_advisory_unlock, lock_id))
    raise "Failed to release advisory lock #{lock_name}"
  end
end

## Then we do:

with_advisory_lock(:migrator) do
  Sequel::Migrator.run DB, 'db/migrations'
end
```

I don't know if this has been considered before (I couldn't find anything about it) but, do you think it would make sense to bring this functionality to the Sequel migrator itself? I'd be willing to prepare a PR if there's appetite for it.

**Comment by jeremyevans** [comment]
Sequel doesn't currently support this, but it does seem worth adding.  I'll definitely consider a pull request that adds this.
